bloomfilter/src/Data/BloomFilter.hs

```diff
@@ -0,0 +1,265 @@
+        -> Bloom' h a
+{-# INLINE create #-}
+create hash numBits body = runST $ do
+    mb <- new hash numBits
+    body mb
+    unsafeFreeze mb
+
+-- | Create an immutable Bloom filter from a mutable one.  The mutable
+-- filter may be modified afterwards.
+freeze :: MBloom' s h a -> ST s (Bloom' h a)
+freeze mb = Bloom (MB.hashesN mb) (MB.size mb) `liftM`
+            V.freeze (MB.bitArray mb)
+
+-- | Create an immutable Bloom filter from a mutable one.  The mutable
+-- filter /must not/ be modified afterwards, or a runtime crash may
+-- occur.  For a safer creation interface, use 'freeze' or 'create'.
+unsafeFreeze :: MBloom' s h a -> ST s (Bloom' h a)
+unsafeFreeze mb = Bloom (MB.hashesN mb) (MB.size mb) `liftM`
+                    V.unsafeFreeze (MB.bitArray mb)
+
+-- | Copy an immutable Bloom filter to create a mutable one.  There is
+-- no non-copying equivalent.
+thaw :: Bloom' h a -> ST s (MBloom' s h a)
+thaw ub = MB.MBloom (hashesN ub) (size ub) `liftM` V.thaw (bitArray ub)
+
+-- | Create an empty Bloom filter.
+empty :: Int                    -- ^ number of hash functions to use
+      -> Word64                 -- ^ number of bits in filter
+      -> Bloom' h a
+{-# INLINE [1] empty #-}
+empty hash numBits = create hash numBits (\_ -> return ())
+
+-- | Create a Bloom filter with a single element.
+singleton :: (Hashes h, Hashable a)
+          => Int               -- ^ number of hash functions to use
+          -> Word64            -- ^ number of bits in filter
+          -> a                 -- ^ element to insert
+          -> Bloom' h a
+singleton hash numBits elt = create hash numBits (\mb -> insert mb elt)
+
+-- | Query an immutable Bloom filter for membership.  If the value is
+-- present, return @True@.  If the value is not present, there is
+-- /still/ some possibility that @True@ will be returned.
+elem :: (Hashes h, Hashable a) => a -> Bloom' h a -> Bool
+elem elt ub = elemHashes (makeHashes elt) ub
+{-# SPECIALIZE elem :: Hashable a => a -> Bloom a -> Bool #-}
+
+-- | Query an immutable Bloom filter for membership using already constructed 'Hashes' value.
+elemHashes :: Hashes h => h a -> Bloom' h a -> Bool
+elemHashes !ch !ub = go 0 where
+    go :: Int -> Bool
+    go !i | i >= hashesN ub = True
+          | otherwise       = let !idx' = evalHashes ch i in
+                              let !idx = idx' `rem` size ub in
+                              if V.unsafeIndex (bitArray ub) idx
+                              then go (i + 1)
+                              else False
+{-# SPECIALIZE elemHashes :: CheapHashes a -> Bloom a -> Bool #-}
+
+-- | Query an immutable Bloom filter for non-membership.  If the value
+-- /is/ present, return @False@.  If the value is not present, there
+-- is /still/ some possibility that @False@ will be returned.
+notElem :: (Hashes h, Hashable a) => a -> Bloom' h a -> Bool
+notElem elt ub = notElemHashes (makeHashes elt) ub
```

Comment:
Isn't this just
```suggestion
notElem elt ub = not (elem elt ub)
```
and so we don't need `notElemHashes` at all?

Reply:
it's whether we trust GHC to do somewhat involved `case of case` transformation here.